fix: unlock callback and buffer for outstanding read when closing port

laqrix opened the pull request:
We typically use a pattern of an async reader process. If that process
is blocked on IO while another process closes the underlying port, the
async reader would not only never return, but would also never be collected.

The osi_read_port callback closed over the process and is locked while
the read is outstanding. Once closed, LibUV will never call its callback
routine which unlocks the Scheme callback and buffer.

The fix unlocks both the Scheme read callback and the read buffer and calls
the read callback with UV_EOF. This allows the Scheme code to process
an end-of-file and any remaining computation including winders.

The order of the read callback and the close callback should not matter.
close-osi-port immediately sets the port-handle to #f, which makes any
subsequent read calls on alternate processes safely fail.

burgerrg commented:
@laqrix, I pushed a commit to fix the automated tests.

laqrix commented:
I believe the case for write_buffer and write_callback is less necessary.  If you are using Scheme's port abstraction, close-port will induce a flush-output-port. If there is a write outstanding, you are likely to have a multi-threaded race problem. If you are in the same process, the write will fully complete before the close.  If we are concerned someone is bypassing the Scheme port abstraction, then I completely agree, but we will need to find a way to test this case at the OSI level.

laqrix commented:
I changed my mind. We should update write_callback and write_buffer for completeness and correctness. I'm not sure how to handle the testing, but we should try something.

burgerrg commented:
Yes, I think it's a good idea to keep it complete & consistent. For testing, try a pipe to another Swish process that doesn't read its input. Fill the pipe to get a write that doesn't complete. Then close the port and watch the sparks fly!

laqrix commented:
Take a look at the `write-leak` mat. Perhaps I missed an obvious path. When I run it, I always get UV_ECANCELED from the write_callback before the close callback fires (at least on Windows and Linux). That matches my expectations at an OS level. Any suggestions to improve the test?

laqrix commented:
@owaddell found the documentation for libuv's [uv_close](http://docs.libuv.org/en/v1.x/handle.html) explicitly states:

> In-progress requests, like uv_connect_t or uv_write_t, are cancelled and have their callbacks called asynchronously with status=UV_ECANCELED.

burgerrg commented:
Since uv_close explicitly guarantees that outstanding write callbacks are invoked before the descriptor is closed, should we simply leave the code as is?

burgerrg commented:
I'm a little surprised that libuv doesn't handle outstanding reads the same way.

laqrix commented:
I was also surprised. I think the asymmetry in the code is unfortunate, but reasonable given the circumstance. Maybe we should add the write-leak test as a regression to make sure the behavior never changes in the future?

burgerrg commented:
Yes, I think it's fine to leave the code asymmetric and to add a write test to verify proper write callback usage when a handle is closed. Thanks!

laqrix commented:
@burgerrg I added a comment to `close_stream_cb` explaining the LibUV behavior (and updated the commit comment). I renamed the tests to `read-leak` and `write-leak`. My second force push also included a check that the winders ran in the write-leak test.

burgerrg commented on src/swish/osi.c:
```diff
@@ -466,9 +466,19 @@ static ptr write_stream_port(uptr port, ptr buffer, size_t start_index, uint32_t
+  if (read_callback) {
+    Sunlock_object(read_buffer);
+    Sunlock_object(read_callback);
+    osi_add_callback1(read_callback, Sfixnum(UV_EOF));
```
Should we use UV_ECANCELED for consistency?

laqrix replied:
I think that might create unexpected behavior from the Scheme port perspective. Where you might have received either a values or an eof-object, a cancel will cause an exception. At least in the cases I've been working through, my primary process has explicitly closed the port, and I expect an eof-object, not an exception.

burgerrg replied:
That seems reasonable.

laqrix commented:
Thanks for the review!
